### PR TITLE
Add unit tests for controller.go

### PR DESCRIPTION
**Pull request by dannawang0221**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug

/kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
Add unit tests for controller.go to make sure code changes is backward compatible.
**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: populator-machinery/controller_test.go [hunk 1]
```diff
@@ -0,0 +1,538 @@
...
+	"k8s.io/client-go/dynamic/dynamiclister"
+	dynamicfake "k8s.io/client-go/dynamic/fake"
+	kubeinformers "k8s.io/client-go/informers"
+	kubefake "k8s.io/client-go/kubernetes/fake"
+	"k8s.io/client-go/util/workqueue"
+	gatewayfake "sigs.k8s.io/gateway-api/pkg/client/clientset/versioned/fake"
+	gatewayInformers "sigs.k8s.io/gateway-api/pkg/client/informers/externalversions"
+)
+
+type testCase struct {
```
Comment (saikat-royc) [inline populator-machinery/controller_test.go:44]: a comment on what each field indicates makes the code more readable

File: populator-machinery/controller_test.go [hunk 2]
```diff
@@ -0,0 +1,538 @@
...
+func TestSyncPvc(t *testing.T) {
+	dataSourceKey := "unstructured/" + testPvcNamespace + "/" + testDataSourceName
+	storageClassKey := "sc/" + testStorageClassName
+	podKey := "pod/" + testVpWorkingNamespace + "/" + testPodName
+	pvcPrimeKey := "pvc/" + testVpWorkingNamespace + "/" + testPopulatorPvcName
+	pvKey := "pv/" + testPvName
+
+	tests := []testCase{
+		{
+			name:         "Ignore PVCs in working namespace",
```
Comment (saikat-royc) [inline populator-machinery/controller_test.go:309]: nit: Ignore PVCs in controller's working namespace

File: populator-machinery/controller_test.go [hunk 3]
```diff
@@ -0,0 +1,538 @@
...
+			pvcNamespace: testPvcNamespace,
+			pvcName:      testPvcName,
+			initialObjects: []runtime.Object{
+				pvc(testPvcName, testPvcNamespace, testNodeName, testStorageClassName, "",
+					dsf(testApiGroup, testDatasourceKind, testDataSourceName, testPvcNamespace), ""),
+				ust(),
+				sc(),
+				pod(corev1.PodSucceeded),
+			},
+			expectedResult: errors.New("Failed to find PVC for populator pod"),
```
Comment (saikat-royc) [inline populator-machinery/controller_test.go:482]: comment for future PRs: We should capture the Pod and PVC name in the error log.

File: populator-machinery/controller_test.go [hunk 4]
```diff
@@ -0,0 +1,538 @@
...
+				"name":      testDataSourceName,
+				"namespace": testPvcNamespace,
+			},
+		},
+	}
+}
+
+func sc() *storagev1.StorageClass {
+	sc := testStorageClassName
+	p := "test.provisioner"
```
Comment (saikat-royc) [inline populator-machinery/controller_test.go:153]: let's use a variable for the provisioner name too?

**Comment by saikat-royc** [comment]
overall looks good. Thanks for adding the unit tests

**Comment by saikat-royc** [comment]
We should also add the unit test to the presubmit run

**Comment by dannawang0221** [comment]
> We should also add the unit test to the presubmit run

Is there an example of doing this?

File: populator-machinery/controller_test.go (hunk 3 above)
Comment (dannawang0221) [reply populator-machinery/controller_test.go:482]: Good point! I want to do this update with a separate PR, and keep this just for adding tests

File: populator-machinery/controller_test.go (hunk 4 above)
Comment (dannawang0221) [reply populator-machinery/controller_test.go]: Done

File: populator-machinery/controller_test.go (hunk 2 above)
Comment (dannawang0221) [reply populator-machinery/controller_test.go]: Done

File: populator-machinery/controller_test.go (hunk 1 above)
Comment (dannawang0221) [reply populator-machinery/controller_test.go:44]: Updated

**Comment by msau42** [comment]
> We should also add the unit test to the presubmit run

I see that it ran:
```
### test-go:
go test -mod=vendor `go list -mod=vendor ./... | grep -v -e 'vendor' -e '/test/e2e$' ` 
?   	github.com/kubernetes-csi/lib-volume-populator/example/hello-populator	[no test files]
?   	github.com/kubernetes-csi/lib-volume-populator/release-tools	[no test files]
ok  	github.com/kubernetes-csi/lib-volume-populator/populator-machinery	3.153s
```

File: populator-machinery/controller_test.go [hunk 5]
```diff
@@ -0,0 +1,546 @@
+/*
+Copyright 2022 The Kubernetes Authors.
```
[the remark this answers is not in the document]
Comment (dannawang0221) [reply populator-machinery/controller_test.go]: Done

File: populator-machinery/controller_test.go [hunk 6]
```diff
@@ -0,0 +1,546 @@
...
+		Group:    testApiGroup,
+		Version:  "v1alpha1",
+		Resource: "testdatasources",
+	}
+	gk = schema.GroupKind{
+		Group: testApiGroup,
+		Kind:  testDatasourceKind,
+	}
+
+	kubeClient    = kubefake.NewSimpleClientset()
```
[the remark this answers is not in the document]
Comment (dannawang0221) [reply populator-machinery/controller_test.go]: Updated

**Pull request merged**